lib/db/transactions.go

```diff
@@ -453,26 +456,28 @@ func (t readWriteTransaction) close() {
 	t.WriteTransaction.Release()
 }
 
-func (t readWriteTransaction) putFile(key []byte, fi protocol.FileInfo) error {
-	if fi.Blocks != nil {
-		if fi.BlocksHash == nil {
-			fi.BlocksHash = protocol.BlocksHash(fi.Blocks)
-		}
-		blocksKey := t.keyer.GenerateBlockListKey(nil, fi.BlocksHash)
-		if _, err := t.Get(blocksKey); backend.IsNotFound(err) {
+func (t readWriteTransaction) putFile(fkey []byte, fi protocol.FileInfo) error {
+	var bkey []byte
+
+	if len(fi.Blocks) > blocksIndirectionCutoff {
+		fi.BlocksHash = protocol.BlocksHash(fi.Blocks)
+		bkey = t.keyer.GenerateBlockListKey(bkey, fi.BlocksHash)
+		if _, err := t.Get(bkey); backend.IsNotFound(err) {
 			// Marshal the block list and save it
 			blocksBs := mustMarshal(&BlockList{Blocks: fi.Blocks})
-			if err := t.Put(blocksKey, blocksBs); err != nil {
+			if err := t.Put(bkey, blocksBs); err != nil {
 				return err
 			}
 		} else if err != nil {
 			return err
 		}
+		fi.Blocks = nil
+	} else {
+		fi.BlocksHash = nil
```

Comment:
I don't like this. I think this is a generally useful piece of information to have. Perhaps we should check if the block list is empty when loading instead to decide if we need to take an indirection or not.

Reply:
Also, avoid adding to bloom filter if it has a blockhash but also a blocklist in that case?

Reply:
We could. You mean that we should always calculate, set and send the hash, even when we don't need it internally?

Reply:
I sort of agree, even though the review comment below suggests the opposite: I feel like `BlocksHash` should either be a db internal implementation detail or always be there if exposed towards protocol.